pkg/controller/atlasdeployment/customzonemapping.go

```diff
@@ -0,0 +1,167 @@
+package atlasdeployment
+
+import (
+	"context"
+	"fmt"
+
+	"go.mongodb.org/atlas/mongodbatlas"
+
+	mdbv1 "github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1/status"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/atlasdeployment/globaldeployment"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/workflow"
+)
+
+func EnsureCustomZoneMapping(service *workflow.Context, groupID string, customZoneMappings []mdbv1.CustomZoneMapping, deploymentName string) workflow.Result {
+	result := syncCustomZoneMapping(context.Background(), service, groupID, deploymentName, customZoneMappings)
+	if !result.IsOk() {
+		service.SetConditionFromResult(status.CustomZoneMappingReadyType, result)
+		return result
+	}
+
+	if customZoneMappings == nil {
+		service.UnsetCondition(status.CustomZoneMappingReadyType)
+		service.EnsureStatusOption(status.AtlasDeploymentCustomZoneMappingOption(nil))
+	} else {
+		service.SetConditionTrue(status.CustomZoneMappingReadyType)
+	}
+
+	return result
+}
+
+func syncCustomZoneMapping(ctx context.Context, service *workflow.Context, groupID string, deploymentName string, customZoneMappings []mdbv1.CustomZoneMapping) workflow.Result {
+	logger := service.Log
+	err := verifyZoneMapping(customZoneMappings)
```

Comment:
K8S unique validation does not support such fields. 
`*spec.validation.openAPIV3Schema.properties[spec].properties[advancedDeploymentSpec].properties[customZoneMapping].uniqueItems: Forbidden: uniqueItems cannot be set to true since the runtime complexity becomes quadratic`

Reply:
I thought about something here https://github.com/mongodb/mongodb-atlas-kubernetes/blob/31df5cf474ef3e21589ef010b7fb049237d919cc/pkg/controller/validate/validate.go#L16